### DISCUSSION TITLE
topological sorting of workspaces when executing commands.

### DISCUSSION THREAD

**Discussion by trusktr** [npm/feedback#579] [CLI] [answered]
It would be nice if we had an option to avoid having to ensure that workspaces in the `workspaces` array are listed in order for cases like running a build step in all packages.

Currently, we must ensure they are in order, so that a dependency of a dependent package will be built before the dependent package get built when running a command like `npm run build --workspaces`.

With a new `topologicalSort: true` option, or similar, we could opt into new behavior that automatically sorts workspaces when running a command that uses `--workspace` or `--workspaces` based on the dependency graph, and executes the commands in the correct order.

Furthermore, having topological sort in place will provide the opportunity for the `npm` cli to be able to run processes of each package in parallel, like Lerna does. F.e. if A depends on B and C, but B and C do not depend on A or on each other, then `npm run build --workspaces` can run `build` for both `B` and `C` in parallel, then run `build` for A once both B and C complete.

**Comment by MylesBorins** [comment, marked as answer]
@trusktr if you would like to suggest new features to the CLI the best course of action is to open up an issue over at https://github.com/npm/rfcs
